### PR TITLE
Fix build issues - macOS rpath, wheel paths, missing pytest

### PR DESCRIPTION
**Pull request by speth**
**Changes proposed in this pull request**

- Fix SCons always rebuilding wheels due to not predicting the file name correctly -- I ended up adopting @bryanwweber's approach for setting `MACOSX_DEPLOYMENT_TARGET` from #1158 after all 
- Force SCons to include `rpath` flags when linking on macOS so libraries that are not in system locations can be used for dependencies (such as libraries installed in a conda environment)
- Provide a better error message in the test summary when `pytest` is not installed
- Suppress XML deprecation warnings from `ctml_writer` that are generated during `scons build`

**Checklist**

- [x] The pull request includes a clear description of this code change
- [x] Commit messages have short titles and reference relevant issues
- [x] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [x] Style & formatting of contributed code follows [contributing guidelines](https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md)
- [x] The pull request is ready for review

### CODE REVIEW

**Review by ischoegl — DISMISSED** [review]

**Comment by bryanwweber** [comment]
Looks like the docs need a pinned jinja2, like the website did

**Comment by bryanwweber** [comment]
The Windows failures are due to Setuptools. Something with Python 3.7 apparently. I'll look when I'm at a computer.

**Comment by speth** [comment]
I'm very confused by the error building the Python module in those two instances (which I don't think has anything to do with the changes introduced in this PR, given `main` is also failing). Searching turned up a couple of other instances reporting failures at the same place inside `distutils`, but no resolution. For example: https://github.com/tensorflow/tensorflow/issues/27477 and https://github.com/jwyang/faster-rcnn.pytorch/issues/681.

For posterity, the traceback in the build log is:
```pytb
  running build_ext
  building 'cantera._cantera' extension
  Traceback (most recent call last):
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 363, in <module>
      main()
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 345, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 262, in build_wheel
      metadata_directory)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\build_meta.py", line 245, in build_wheel
      wheel_directory, config_settings)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\build_meta.py", line 229, in _build_with_temp_dir
      self.run_setup()
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\build_meta.py", line 174, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 5, in <module>
      setup(ext_modules=[extension])
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\__init__.py", line 79, in setup
      return distutils.core.setup(**attrs)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\core.py", line 148, in setup
      return run_commands(dist)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\core.py", line 163, in run_commands
      dist.run_commands()
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\dist.py", line 967, in run_commands
      self.run_command(cmd)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\dist.py", line 1196, in run_command
      super().run_command(command)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\dist.py", line 986, in run_command
      cmd_obj.run()
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\wheel\bdist_wheel.py", line 299, in run
      self.run_command('build')
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\dist.py", line 1196, in run_command
      super().run_command(command)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\dist.py", line 986, in run_command
      cmd_obj.run()
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\command\build.py", line 135, in run
      self.run_command(cmd_name)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\dist.py", line 1196, in run_command
      super().run_command(command)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\dist.py", line 986, in run_command
      cmd_obj.run()
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\command\build_ext.py", line 79, in run
      _build_ext.run(self)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\Cython\Distutils\old_build_ext.py", line 186, in run
      _build_ext.build_ext.run(self)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\command\build_ext.py", line 339, in run
      self.build_extensions()
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\Cython\Distutils\old_build_ext.py", line 195, in build_extensions
      _build_ext.build_ext.build_extensions(self)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\command\build_ext.py", line 448, in build_extensions
      self._build_extensions_serial()
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\command\build_ext.py", line 473, in _build_extensions_serial
      self.build_extension(ext)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\command\build_ext.py", line 202, in build_extension
      _build_ext.build_extension(self, ext)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\command\build_ext.py", line 559, in build_extension
      target_lang=language)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\ccompiler.py", line 717, in link_shared_object
      extra_preargs, extra_postargs, build_temp, target_lang)
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\setuptools\_distutils\_msvccompiler.py", line 483, in link
      build_temp = os.path.dirname(objects[0])
  IndexError: list index out of range
  [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for Cantera
Failed to build Cantera
ERROR: Failed to build one or more wheels
scons: *** [build\python\dist\Cantera-2.6.0b1-cp37-cp37-win_amd64.whl] Error 1
```

**Comment by speth** [comment]
Investigating the more detailed logging output that has been temporarily enabled, the difference that I've noticed between these two cases is that when the build succeeds, the logs for `pip wheel` contain (in part):
```
  ...
  creating build\lib.win-amd64-3.9\cantera\examples\thermo
  creating build\lib.win-amd64-3.9\cantera\examples\transport
  running build_ext
  installing to build\bdist.win-amd64\wheel
  running install
  running install_lib
  ...
```
while for the failing builds, they contain the following:
```
  creating build\lib.win-amd64-3.7\cantera\examples\thermo
  creating build\lib.win-amd64-3.7\cantera\examples\transport
  running build_ext
  building 'cantera._cantera' extension
  Traceback (most recent call last):
    File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py", line 363, in <module>
      main()
```
which indicates that in the failing case, setuptools or wheel or pip (I really don't understand any of this infrastructure) erroneously thinks there's actually something that needs to be compiled.

**Comment by speth** [comment]
Looking through earlier builds on the `main` branch, the failing builds with Python 3.7 are due to changing from `setuptools` 60.x to 61.x. Pinning `setuptools` to version 60 on those builders seems to fix the problem.

**Comment by speth** [comment]
The [release notes for `setuptools` 61.0.0](https://setuptools.pypa.io/en/latest/history.html#v61-0-0) mentions https://github.com/pypa/setuptools/pull/2894 as a "breaking change"
> Projects that currently don’t specify both packages and py_modules in their configuration and contain extra folders or Python files (not meant for distribution), might see these files being included in the wheel archive or even experience the build to fail.

If this is the cause of our problems, I have no idea why it's restricted to such a narrow configuration of builders. Also, I find it irritating that there is no logging information even at the highest debug loglevel about what `setuptools` is trying to do during the `build_ext` step.

File: interfaces/cython/setup-debian.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# This script is only used with on Debian based systems with the patched
+# versions of patched versions of setuptools that are installed there.
```
Comment (ischoegl) [inline interfaces/cython/setup-debian.py]: patched versions of patched versions ... :joy:

**Review by ischoegl — COMMENTED** [review]
Don't have much to add here.

File: interfaces/cython/setup-debian.py (hunk 1 above)
Comment (speth) [reply interfaces/cython/setup-debian.py]: Well, it's probably true, but yes. 😂

Comment (ischoegl) [reply interfaces/cython/setup-debian.py]: honestly thought this was intentional ...

**Comment by bryanwweber** [comment]
Thanks for all the work here @speth! I'll review it tonight. My only concern right away is that the wheel has the right name, it should have some platform information in it.

**Comment by speth** [comment]
Huh, I didn't realize that the bits in `setup.py` were what was responsible for tricking `setuptools` into generating a platform-specific wheels. Let's see if overriding `setuptools.dist.Distribution.has_ext_modules` works any better...

**Comment by bryanwweber** [comment]
> Huh, I didn't realize that the bits in `setup.py` were what was responsible for tricking `setuptools` into generating a platform-specific wheels.

It's not a trick, really... C extensions are what cause the requirement for platform-specific wheels, so you have to tell setuptools that it's building an extension. Previously, it decided that since there were no source files, it could just package the library. Dunno why that stopped working.

**Comment by speth** [comment]
Yes, I understand that the compiled extension is the reason that the wheel *should* be platform specific. But what was being specified was an empty extension module, while we just stick the actual compiled extension module in as if it's just "data" and `setuptools` does not recognize it as creating a platform dependency (which would be asking a lot). So I think this "empty extension module" really is a bit of a hack and setuptools ought to provide a better way of handling situations where there is platform-specific package content but it's not being compiled by `setuptools`.

File: interfaces/cython/setup.cfg.in [hunk 2]
```diff
@@ -49,6 +49,7 @@ packages =
     cantera.test
     cantera.test.data
     cantera.examples
+py_modules =
```
Comment (bryanwweber) [inline interfaces/cython/setup.cfg.in]: I don't think this line is required. At least, the same files are packaged in the wheel whether this is here or not.

File: test/python/runCythonTests.py [hunk 3]
```diff
@@ -28,19 +28,16 @@
...
-
-
-if __name__ == "__main__":
-    if pytest is None:
         print("\n* ERROR: The Cantera Python test suite requires "
             "the Python package 'pytest'.")
         print("* ERROR: Use pip or conda to install 'pytest', "
             "which will enable this feature.")
-        sys.exit(1)
+        sys.exit(21)
```
Comment (bryanwweber) [inline test/python/runCythonTests.py]: Was this change intentional?

**Review by bryanwweber — DISMISSED** [review]
Thanks @speth! The tests pass and I can confirm that I'm getting platform specific wheels locally. Just a couple minor comments here.

File: test/python/runCythonTests.py (hunk 3 above)
Comment (speth) [reply test/python/runCythonTests.py]: Yes, this specific error code now checked for in `test/SConscript` so we can provide a better error message in the case where `pytest` is not installed rather than just "the entire Python test suite failed". I'll add a comment for this.

File: interfaces/cython/setup.cfg.in (hunk 2 above)
Comment (speth) [reply interfaces/cython/setup.cfg.in]: This was based on reading of the (conflicting) documentation about this new "discovery" feature of `setuptools`, where it says:
> Projects that currently don’t specify both packages and py_modules in their configuration and contain extra folders or Python files (not meant for distribution), might see these files being included in the wheel archive or even experience the build to fail.

Elsewhere, it says:
> Automatic discovery will only be enabled if you don’t provide any configuration for packages and py_modules. If at least one of them is explicitly set, automatic discovery will not take place.

So I really don't know. But I guess we can see if everything still works without it.

Comment (bryanwweber) [reply interfaces/cython/setup.cfg.in]: I have setuptools 61.1.1, so I'm not sure.

Their documentation is _terrible_, a mishmash of stuff copied from the original distutils (and still duplicated in the official Python docs), random pages of (maybe?) best practices, and no complete API specification. But hey, it looks pretty!

File: interfaces/cython/setup.py [hunk 4]
```diff
@@ -1,5 +1,10 @@
-from setuptools import setup, Extension
+from setuptools import setup
+from setuptools.dist import Distribution
 
-extension = Extension("cantera._cantera", sources=[])
+class BinaryDistribution(Distribution):
+    def has_ext_modules(self):
+        return True
```
Comment (bryanwweber) [inline interfaces/cython/setup.py:6]: This is my new favorite Python class ❤️

**Review by bryanwweber — APPROVED** [review]
Thanks again @speth, let's :shipit:
